### PR TITLE
First draft for educational environments

### PR DESCRIPTION
**Pull request by hannahpotter**

### CODE REVIEW

File: src/educational-technology.rst [hunk 1]
```diff
@@ -1,18 +1,183 @@
...
+.. :Authors: - Cyrus Omar, Hannah Potter
 
 .. title:: Educational Technology
 
 Overview
 ========
 
-TODO: Overview
+Educational technology is primarily focused at facilitating learning. This includes supporting novice programmers learning the fundamentals
+of programming, supplying tools that promote knowledge sharing, assisting educators in their teaching objectives, and
```
Comment (cyrus-) [inline src/educational-technology.rst]: rephrase this so that every element of this list takes the form (user group) (activity), like "novice programmers learning the fundamentals of programming"

File: src/educational-technology.rst [hunk 2]
```diff
@@ -1,18 +1,183 @@
-.. :Authors: - Cyrus Omar
+.. :Authors: - Cyrus Omar, Hannah Potter
 
 .. title:: Educational Technology
 
 Overview
 ========
 
-TODO: Overview
+Educational technology is primarily focused at facilitating learning. This includes supporting novice programmers learning the fundamentals
```
Comment (cyrus-) [inline src/educational-technology.rst]: why "primarily"?

File: src/educational-technology.rst [hunk 3]
```diff
@@ -1,18 +1,183 @@
...
+of programming, supplying tools that promote knowledge sharing, assisting educators in their teaching objectives, and 
+facilitating learning of new technologies and systems. 
+
+Programming environments that are being used during a learning process focus on different design needs than those that are
```
Comment (cyrus-) [inline src/educational-technology.rst]: "that are being used during" -> "designed for use during"
"focus on different design needs" -> "are governed by different design criteria"

File: src/educational-technology.rst [hunk 4]
```diff
@@ -1,18 +1,183 @@
...
+designed to be useful for experts. Educational programming environments typically focus on trying to mitigate what people typically
```
Comment (cyrus-) [inline src/educational-technology.rst]: The two points in this sentence seem distinct, so they should be separate sentences

File: src/educational-technology.rst [hunk 5]
```diff
@@ -1,18 +1,183 @@
...
+find to be hard about programming while supporting the motivations that different people have for trying to learn these new skills :cite:`guzdial2004programming` :cite:`kelleher2005lowering`.
+A few general strategies that are employed to accomplish these goals are developing special programming languages for learning and developing special
```
Comment (cyrus-) [inline src/educational-technology.rst]: This point should probably be made at the beginning of this paragraph, e.g. "Programming languages and environments that are being designed ..."

File: src/educational-technology.rst (hunk 4 above)
Comment (cyrus-) [reply src/educational-technology.rst]: instead of "mitigate what people typically find to be hard about programming", think about rephrasing this in terms of "learning curves", e.g. the focus is on flattening the learning curve

File: src/educational-technology.rst [hunk 6]
```diff
@@ -1,18 +1,183 @@
...
+environments (without changing the programming language).
+
+There are a few key design principles that are often used in either of these strategies. One is to try to minimize frustration while still being
+challenging enough to maintain interest, with the related principle of incorporating motivations for learning these skills. 
+Another is related to being as concrete as possible versus abstract, as people need to understand concrete concepts before they
+can begin to apply abstractions. Additionally, educational programming environments often are designed to give immediate feedback, often 
+in a visual way that is easy for a learner to understand and see what is happening.
```
Comment (cyrus-) [inline src/educational-technology.rst]: "in a visual way" -> "using visual notation"
"that is easy for a learner to understand and see what is happening" -> "that helps the user confirm or refine their beliefs about how the program being written behaves"

File: src/educational-technology.rst [hunk 7]
```diff
@@ -1,18 +1,183 @@
...
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'maloney2010scratch'
+
+  This paper gives an overview of the design of the Scratch programming environment. Scratch is a graphical, block-based programming environment
+  designed for kids who have no programming experience. The system strives to make execution visible and allow for tinkering (learning by trying). The
+  code is live such that any code fragment can be executed simply by clicking on it (no compile-link-run cycle to go through).
+
+.. todo:: 
+    Find evaluation of effectiveness of Scratch
```
Comment (cyrus-) [inline src/educational-technology.rst]: Finding a paper or two evaluating Scratch would be useful for your second draft

File: src/educational-technology.rst [hunk 8]
```diff
@@ -1,18 +1,183 @@
...
+For a more formal classroom setting, teachers can often use support in getting a better understanding of what topics
+their students are excelling at understanding and what topics their students are struggling to understand.
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'glassman2015overcode'
+
+  OverCode is a system designed to allow instructors of large programming classes to automatically group solutions that may have 
+  irrelavent syntactic differences but are semantically equivalent. This should allow instructors to get a high-level overview of
```
Comment (cyrus-) [inline src/educational-technology.rst]: "irrelavent" -> "irrelevant"

File: src/educational-technology.rst [hunk 9]
```diff
@@ -1,18 +1,183 @@
...
+  the understandings and misunderstandings of their students more quickly than filtering through raw solutions. When evaluated against
+  only having access to raw solutions, users were able to review more students' code in less time and felt that they had a better high-level
+  overview of ways that the coding problems were solved. 
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'head2018interactive'
+
+  CodeScoop is an interactive example extraction system that allows users to quickly pull out examples from an existing code
```
Comment (cyrus-) [inline src/educational-technology.rst:214]: You should justify why this is under "Educator Support" (e.g. by mentioning that the focus of the paper is on answering questions on Stack Overflow)

File: src/educational-technology.rst [hunk 10]
```diff
@@ -1,18 +1,183 @@
...
+    :filter: key == 'guo2013online'
+
+  Python Tutor is an online tutoring system. This is a form of program visualization that shows the state of memory (stack frames and the heap)
+  as a piece of code executes, essentially creating visual code traces. 
+
+.. todo::
+    Add information about systems that introduce the programming language a bit at a time (like SP/k)
+
+.. todo::
+    Find interactive tutoring systems
```
Comment (cyrus-) [inline src/educational-technology.rst]: Look a little more into tutoring systems for your second draft

**Review by cyrus- — CHANGES_REQUESTED** [review]
looks good overall! see the comments for some changes that you should consider for your second draft

(first draft is only graded for effort, so you passed)

File: src/educational-technology.rst (hunk 7 above)
Comment (cyrus-) [reply src/educational-technology.rst]: here is some work you might refer to here: https://computinged.wordpress.com/2019/08/19/summarizing-findings-about-block-based-programming-in-computing-education/

File: src/educational-technology.bib [hunk 11]
```diff
@@ -0,0 +1,86 @@
+@article{guzdial2004programming,
+  title={Programming environments for novices},
+  author={Guzdial, Mark},
+  journal={Computer science education research},
```
Comment (cyrus-) [inline src/educational-technology.bib]: capitalize

File: src/educational-technology.bib [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  volume={37},
+  number={2},
+  pages={83--137},
+  year={2005},
+  publisher={ACM}
+}
+
+@article{wyeth2001electronic,
+  title={Electronic blocks: Tangible programming elements for preschoolers},
+  journal={},
```
Comment (cyrus-) [inline src/educational-technology.bib:23]: what is the venue?

File: src/educational-technology.bib [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+  number={4},
+  pages={16},
+  year={2010},
+  publisher={ACM}
+}
+
+@inproceedings{guo2013online,
+  title={Online python tutor: embeddable web-based program visualization for cs education},
+  author={Guo, Philip J},
+  booktitle={Proceeding of the 44th ACM technical symposium on Computer science education},
```
Comment (cyrus-) [inline src/educational-technology.bib]: capitalize

File: src/educational-technology.bib [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  author={Fessakis, Georgios and Gouli, Evangelia and Mavroudi, E},
+  journal={Computers \& Education},
+  volume={63},
+  pages={87--97},
+  year={2013},
+  publisher={Elsevier}
+}
+
+@article{maloney2010scratch,
+  title={The scratch programming language and environment},
```
Comment (cyrus-) [inline src/educational-technology.bib]: capitalize Scratch (need curly braces in BibTeX: `{Scratch}` )

File: src/educational-technology.bib [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+  journal={ACM Transactions on Computing Education (TOCE)},
+  volume={10},
+  number={4},
+  pages={16},
+  year={2010},
+  publisher={ACM}
+}
+
+@inproceedings{guo2013online,
+  title={Online python tutor: embeddable web-based program visualization for cs education},
```
Comment (cyrus-) [inline src/educational-technology.bib]: capitalize Python and CS

File: src/educational-technology.bib [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  booktitle={Proceeding of the 44th ACM technical symposium on Computer science education},
+  pages={579--584},
+  year={2013},
+  organization={ACM}
+}
+
+@inproceedings{suzuki2017exploring,
+  title={Exploring the design space of automatically synthesized hints for introductory programming assignments},
+  author={Suzuki, Ryo and Soares, Gustavo and Glassman, Elena and Head, Andrew and D'Antoni, Loris and Hartmann, Bj{\"o}rn},
+  booktitle={Proceedings of the 2017 CHI Conference Extended Abstracts on Human Factors in Computing Systems},
```
Comment (cyrus-) [inline src/educational-technology.bib:62]: is there a full paper about this work?

File: src/educational-technology2.bib [hunk 17]
```diff
@@ -0,0 +1,36 @@
+@book{maloney2008programming,
+  title={Programming by choice: urban youth learning programming with scratch},
+  author={Maloney, John H and Peppler, Kylie and Kafai, Yasmin and Resnick, Mitchel and Rusk, Natalie},
```
Comment (cyrus-) [inline src/educational-technology2.bib]: what is the venue?

File: src/educational-technology2.bib [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  volume={40},
+  number={1},
+  year={2008},
+  publisher={ACM}
+}
+
+@inproceedings{wilson2010evaluating,
+  title={Evaluating Scratch to Introduce Younger Schoolchildren to Programming.},
```
Comment (cyrus-) [inline src/educational-technology2.bib]: No period at the end of titles

File: src/educational-technology2.bib [hunk 19]
```diff
@@ -0,0 +1,36 @@
+@book{maloney2008programming,
+  title={Programming by choice: urban youth learning programming with scratch},
```
Comment (cyrus-) [inline src/educational-technology2.bib]: capitalize Scratch

File: src/educational-technology2.bib [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  number={3},
+  pages={404--415},
+  year={2015},
+  publisher={IEEE}
+}
+
+@inproceedings{gerdes2012interactive,
+  title={An interactive functional programming tutor},
+  author={Gerdes, Alex and Jeuring, Johan and Heeren, Bastiaan},
+  booktitle={Proceedings of the 17th ACM annual conference on Innovation and technology in computer science education},
```
Comment (cyrus-) [inline src/educational-technology2.bib]: Capitalize correctly

File: src/educational-technology.rst [hunk 21]
```diff
@@ -1,18 +1,227 @@
...
+
+Programming languages and environments that are designed for use during a learning process are governed by different design criteria than those that are
+designed to be useful for experts. Educational programming environments typically focus on two problems. One is trying to flatten the 
+steep learning curve often faced by novice programmers and help them overcome what is typically perceived as "hard" with 
+programming :cite:`guzdial2004programming`. The second is supporting the motivations that different 
+people have for trying to learn these new programming skills :cite:`kelleher2005lowering`.
+
+There are a few key design principles that are often used in either of these strategies. One is to try to minimize frustration while still being
+challenging enough to maintain interest, with the related principle of incorporating motivations for learning these skills. 
+Another is related to being as concrete as possible versus abstract, as people need to understand concrete concepts before they
```
Comment (cyrus-) [inline src/educational-technology.rst]: instead of making this out to be a "versus" relationship, consider rephrasing to focus on the need to learn abstract concepts using concrete instances

File: src/educational-technology.rst [hunk 22]
```diff
@@ -1,18 +1,227 @@
...
+can begin to apply abstractions. Additionally, educational programming environments often are designed to give immediate feedback,
```
Comment (cyrus-) [inline src/educational-technology.rst]: link to the Live Programming chapter when mentioning immediate feedback

File: src/educational-technology.rst [hunk 23]
```diff
@@ -1,18 +1,227 @@
...
+of programming, supplying tools for developers that promote knowledge sharing, assisting educators in their teaching objectives, and 
+facilitating learning for experienced programmers using new technologies and systems. 
+
+Programming languages and environments that are designed for use during a learning process are governed by different design criteria than those that are
+designed to be useful for experts. Educational programming environments typically focus on two problems. One is trying to flatten the 
+steep learning curve often faced by novice programmers and help them overcome what is typically perceived as "hard" with 
+programming :cite:`guzdial2004programming`. The second is supporting the motivations that different 
+people have for trying to learn these new programming skills :cite:`kelleher2005lowering`.
+
+There are a few key design principles that are often used in either of these strategies. One is to try to minimize frustration while still being
```
Comment (cyrus-) [inline src/educational-technology.rst]: Consider making this a bulleted list instead of a paragraph

File: src/educational-technology.rst [hunk 24]
```diff
@@ -1,18 +1,227 @@
...
+with one way of accomplishing this being using visual notation that helps the user confirm or refine their beliefs about how the
+program being written behaves.
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'guzdial2004programming'
+
+  This paper gives an overview of a variety of different programming environments for novices. The educational environments are explored
+  under the lense of how they try to answer the question of "What makes programming hard?"
```
Comment (cyrus-) [inline src/educational-technology.rst:31]: can you summarize how the paper answers this question?

File: src/educational-technology.rst [hunk 25]
```diff
@@ -1,18 +1,227 @@
...
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'kelleher2005lowering'
+
+  This paper also gives an overview of various programming languages and environments that were designed for supporting novice programmers. 
+  These languages and environments are explored under the lense of supporting the different answers to "What is a person's motivation for trying
```
Comment (cyrus-) [inline src/educational-technology.rst:39]: same, can you summarize how this paper answers this question?

File: src/educational-technology.rst [hunk 26]
```diff
@@ -1,18 +1,227 @@
...
+.. todo:: 
+    Add other game based programming environments
 
 Block-Based Environments
 ========================
 
+Block-based environments are often used for teaching young students (pre-university) and students who have had little or no previous exposure to programming.
+These environments have the advantage that there is a reduced coginitive load on learners when it comes to syntax. Block-based environments can limit or eliminate syntax errors
+which saves novices time and confusion resolving such errors and allows them to focus on the higher level concepts. Additionally, the use of blocks
+minimizes the demand for typing and shows program construction in a graphical way, which is especially helpful for children.
```
Comment (cyrus-) [inline src/educational-technology.rst:68]: you might also want to mention that blocks are familiar to children from other forms of play, e.g. Legos

File: src/educational-technology.rst [hunk 27]
```diff
@@ -1,18 +1,227 @@
...
+  Scratch was used for teaching lessons in an IT class for 8-9 year old students at a school. The study
+  did not find evidence of whether or not using this programming environment helped the students develop cognitive abilities. 
+  However, based on feedback from the students, they generally enjoyed and were enthusiastic about their lessons 
+  that used Scratch. 
+
 Feedback Generation
 ===================
 
+With the growing number of people interested in learning programming skills, institutions have difficulty maintaining the number of instructors
+needed to give students valuable one-on-one feedback. Thus, automatic feedback generation is innvaluable in the current learning climate.
```
Comment (cyrus-) [inline src/educational-technology.rst]: "is innvaluable" -> "may be valuable"

File: src/educational-technology.rst [hunk 28]
```diff
@@ -1,18 +1,227 @@
...
+students can upload submissions and get information back on how their solutions performed against test suites. This however still leaves a burden
+on students to search out where their solution went wrong, which may be difficult for novices. Thus, feedback that can guide students through
+debugging code with errors can be far more valuable to the learning process than what is returned by standard autograders.
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'suzuki2017exploring'
+
+  Five types of common hints that teachers give to help students fix their code that can be generated using program synthesis
```
Comment (cyrus-) [inline src/educational-technology.rst]: what method(s) do they use to identify these things?

File: src/educational-technology.rst [hunk 29]
```diff
@@ -1,18 +1,227 @@
...
+  are itentified: transformations (what to change to make the program work), locations (the line(s) that need to be changed to make the program work),
```
Comment (cyrus-) [inline src/educational-technology.rst]: consider making these numbered lists rather than leaving them inline in the sentence

File: src/educational-technology.rst [hunk 30]
```diff
@@ -1,18 +1,227 @@
...
+of code. Additionally, they may guide students through developing a program and improving problem solving skills. Students may remain more engaged with the system
+and learn more if there is interaction that demonstrates whether or not the student is understanding and following along.
+
+.. container:: bib-item
+
+  .. bibliography:: educational-technology.bib
+    :filter: key == 'guo2013online'
+
+  Python Tutor is an online tutoring system. This is a form of program visualization that shows the state of memory (stack frames and the heap)
+  as a piece of code executes, essentially creating visual code traces.
```
Comment (cyrus-) [inline src/educational-technology.rst]: summarize the evaluation (even if there wasn't one)

File: src/educational-technology.rst [hunk 31]
```diff
@@ -1,18 +1,227 @@
...
+    :filter: key == 'head2018interactive'
+
+  CodeScoop is an interactive example extraction system that allows users to quickly pull out examples from an existing code
+  base. Given selected lines to be included in an example, the system iteratively checks for ommitted
+  lines of code that may be valuable (such as control flow) to help the example writer not leave out critical information. 
+  Study participants were asked to create an example from existing code to answer a fake Stack Overflow question. More participants
+  in the study were able to complete this example extraction exercise when using CodeScoop compared to a text editor and liked their end result
+  example better.
+
+Domain Specific Environments
```
Comment (cyrus-) [inline src/educational-technology.rst]: "Domain-Specific and Task-Specific Environments"

**Review by cyrus- — CHANGES_REQUESTED** [review]
nicely done -- grade and further comments will be sent by email

File: src/educational-technology.bib (hunk 12 above)
Comment (hannahpotter) [reply src/educational-technology.bib:23]: I wasn't able to find a venue it was published in

File: src/educational-technology.bib (hunk 16 above)
Comment (hannahpotter) [reply src/educational-technology.bib:62]: Not that I have been able to find

File: src/educational-technology.bib (hunk 12 above)
Comment (cyrus-) [reply src/educational-technology.bib:23]: Proceedings of the Eighth IFIP TC13 Conference on Human-Computer Interaction (https://www.researchgate.net/publication/43462811_Electronic_blocks_Tangible_programming_elements_for_preschoolers/citation/download)

File: src/educational-technology2.bib [hunk 32]
```diff
@@ -0,0 +1,38 @@
+@inproceedings{maloney2008programming,
+ author = {Maloney, John H. and Peppler, Kylie and Kafai, Yasmin and Resnick, Mitchel and Rusk, Natalie},
+ title = {Programming by Choice: Urban Youth Learning Programming with Scratch},
```
Comment (cyrus-) [inline src/educational-technology2.bib:3]: Need curly braces around `{Scratch}` because it is a proper noun

File: src/educational-technology2.bib [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+ booktitle = {Proceedings of the 39th SIGCSE Technical Symposium on Computer Science Education},
+ series = {SIGCSE '08},
+ year = {2008},
+ pages = {367--371},
+ numpages = {5},
+ publisher = {ACM},
+} 
+
+@inproceedings{wilson2010evaluating,
+  title={Evaluating Scratch to Introduce Younger Schoolchildren to Programming},
```
Comment (cyrus-) [inline src/educational-technology2.bib:13]: same here

**Review by cyrus- — APPROVED** [review]
great! will send final grades via email soon
